### DISCUSSION TITLE
How to use Dependency Injection in GLSP?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#677] [Q&A] [answered]
Hi,

I am trying to use Guice Dependencies in my GLSP Server module. But nothing seems to work. My first thought was that this should be super easy.  

First I declare an Interface

```java
public interface BPMNExtensionHandler {
    String getUISchema(BPMNBaseElement element);
}
```

Next I do some implementation of that interface:

```java
public class BPMNEventExtensionHandler implements BPMNExtensionHandler {

    public BPMNEventExtensionHandler() {
        super();
    }
    @Override
    public String getUISchema(final BPMNBaseElement element) {
        return null;
    }
}
```

Now I can inject this handler via my interface somewere in my code:

```java
public class EventNodeBuilder extends AbstractGNodeBuilder<EventNode, EventNodeBuilder> {

    @Inject
    protected BPMNExtensionHandler bpmnExtension;

    @Override
    public void setProperties(final EventNode node) {
        .....
        // call extension.....
        if (bpmnExtension != null) {
            System.out.println("-------_> go");
        }
    }
}
```

and finally I need to bind my implementation in my DiagramModule. I tried everything, but nothing works....

```java
public class BPMNDiagramModule extends DiagramModule {
    @Override
    protected void configureBase() {
        super.configureBase();

        // bind(BPMNEventExtensionHandler.class).in(Singleton.class);  // does not work
        // does even not work
        bind(BPMNExtensionHandler.class).to(BPMNEventExtensionHandler.class);
    }
}
```

My injected BPMNExtensionHandler is always null. Also the constructor is never called. 

What did I wrong?

I notice that there are tons of code in GLSP providing some kind of helper methods for using Guice dependencies - registries, binding methods, abstract classes, helper classes.....

But it is impossible for me to guess what all this code is good for. 

For example - what is the class `org.eclipse.glsp.server.di.MultiBinding<T>`  good for and how can I reuse it - and should I reuse this class after all?

Can someone provide me an example? 
Actually, I wanted to implement a MultiBinding, but this seems to me to be far away.....

**Comment by CamilleLetavernier** [comment, marked as answer]
Hi Ralph,

The dependency-injection creates a graph of objects. `@Inject` can only be used inside of objects that were also created via dependency-injection. So I think the first question is: how is your EventNodeBuilder created?

**Reply by rsoika** [reply]
Thanks! This was exactly the problem. I was not aware of this important point.
